lib/translator/inquiry.cc

```diff
@@ -0,0 +1,253 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "inquiry.h"
+#include "string.h"
+
+namespace translator {
+
+void TranslateStandardInquiry(
+    const nvme_defs::IdentifyControllerData &identify_controller_data,
+    const nvme_defs::IdentifyNamespace &identify_namespace_data,
+    absl::Span<uint8_t> buffer) {
+  scsi_defs::InquiryData result = {
+      .version = scsi_defs::Version::kSpc4,
+      .response_data_format = scsi_defs::ResponseDataFormat::kCompliant,
+      .additional_length = 0x1f,
+      .tpgs = scsi_defs::TPGS::kNotSupported,
+      .protect = (identify_namespace_data.dps.pit == 0 &&
+                  identify_namespace_data.dps.md_start == 0)
+                     ? 0
+                     : 1,
+      .cmdque = 1};
+
+  // Shall be set to “NVMe" followed by 4 spaces: “NVMe    “
+  // Vendor Identification is not null terminated.
+  static_assert(sizeof(result.vendor_identification) ==
+                kNvmeVendorIdentification.size());
+
+  memcpy(result.vendor_identification, kNvmeVendorIdentification.data(),
+         kNvmeVendorIdentification.size());
+
+  // Shall be set to the first 16 bytes of the Model Number (MN) field within
+  // the Identify Controller Data Structure
+  memcpy(result.product_identification, identify_controller_data.mn,
+         sizeof(result.product_identification));
+
+  /*
+  Shall be set to the last 4 ASCII graphic characters in the range of 21h-7Eh
+  (i.e. last 4 non-space characters) of the Firmware Revision (FR) field within
+  the Identify Controller Data Structure.
+  */
+  int idx = 3;
+  for (int i = 7; i >= 0; i--) {
```

Comment:
nit: c++ style guide advocates pre-decrement and increment when post and pre are equivalent https://google.github.io/styleguide/cppguide.html#Preincrement_and_Predecrement

Reply:
will change but i believe compiler optimizes this